Feat/Record licensee id in privilege transactions

landonshumway-ia opened the pull request:
For reporting purposes, we must capture the user id of the licensee in the transaction record in authorize.net. Our user ids are generated as UUIDv4 strings, which are 36 characters long and contain dashes. Originally, we had intended to use the customer id field authorize.net provides for storing these. However, that field is limited to 20 alphanumeric characters (no dashes). The only field that will support the length and dashes of our user ids is the order description field. This field is readonly once the transaction is created, allows up to 255 characters, and is not restricted to alphanumeric characters. 

This PR updates the Purchase client to pass in the licensee id into the order description, so it can be referenced by our reporting which will be implemented in a future ticket.

### Requirements List
- no new requirements. Using existing information.

### Description List
- Add required 'licensee_id' field when processing privilege purchases.

### Testing List
- `yarn test:unit:all` should run without errors or warnings
- `yarn serve` should run without errors or warnings
- `yarn build` should run without errors or warnings
- For API configuration changes: CDK tests added/updated in `backend/compact-connect/tests/unit/test_api.py`
- Code review

Closes #287

jlkravitz commented on backend/compact-connect/lambdas/purchases/purchase_client.py:
```diff
@@ -208,6 +211,12 @@ def process_charge_on_credit_card_for_privilege_purchase(
+        order = apicontractsv1.orderType()
+        # We store the LICENSEE ID in the description field, since the ID is a UUID that is 36 characters long
+        # and this is the only field that can store that length of data
+        order.description = f'LICENSEE#{licensee_id}#'
```
Is this extensible if we decide to add more data to the description field in the future? What are the tradeoffs of doing it this way, vs. as json (e.g.)?

landonshumway-ia replied:
That's a great question and I think the main answer is character count. We can only store 255 chars in this field, and to support json we would need double quotes around the keys and string values, which adds an extra 4 chars per field. We would also need to track brackets. This current setup stores 46 chars, which leaves room for 209 more chars, or approximately 4-5 more fields. We don't anticipate needing to store any more fields at this point, since the licensee Id is enough to fetch all the related records for the licensee on our side of the system.

landonshumway-ia replied:
However, doing the math, if we were to convert this to into a json formatted string it would put us at 52 characters. That leaves us with 203 characters, which is arguable still plenty of headroom for a couple more variables if needed. I think I prefer formatting it as a json string, as it makes it more clear that this is a data field used by an automation system. With that in mind, I'm going to update this to be a json formatted string.

landonshumway-ia replied:
@jlkravitz so I just tested making a transaction using a json serialized string, and authorize.net removed the brackets around the object. That was unexpected behavior which is not documented in their API. 
![Screenshot 2024-12-05 at 12 29 51 PM](https://github.com/user-attachments/assets/a1610c57-68c1-4b77-b3c1-65f676cd256a)

Obviously, this will not work, so I'm going to revert back to using the # delimiter.

jlkravitz replied:
Strange! But sounds good.

jlkravitz reviewed (commented):
One question, but otherwise looks good!

landonshumway-ia commented:
I've rebased the branch with development and updated our API spec to latest, hence the large diff. The actual code changes are still relatively small.

jlkravitz reviewed (approved):
Looks great! Thanks for adding the comment about our consideration of JSON. Helpful for our future brains.

@isabeleliassen This is good to merge!
